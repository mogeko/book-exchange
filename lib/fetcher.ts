async function fetcher(url: string, init?: RequestInit): Promise<Response> {
  const res = await fetch(url, init);

  if (!res.ok) {
    const error = new NetworkError("No Content");
    error.info = await res.json();
    error.status = res.status;
    throw error;
  }

  return res.json();
}

export class NetworkError extends Error {
  constructor(message?: string) {
    super(message);
    this.name = "NetworkError";
  }
  status?: number;
  info?: any;
}

export default fetcher;
