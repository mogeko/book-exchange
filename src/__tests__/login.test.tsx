import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import fetcher from "@/lib/fetcher";

describe("Authorize API", () => {
  it("return a hmac salt", async () => {
    const { salt } = await fetcher<Salt>("/api/auth/salt");

    expect(salt).toBeDefined();

    expect(salt).toMatchSnapshot();
  });

  it("return an authorization token", async () => {
    const auth = await fetcher<ResType>("/api/auth/login", {
      method: "POST",
      body: JSON.stringify({
        username: "test",
        password: "test",
      }),
      headers: {
        "Content-Type": "application/json",
      },
    });
    const getCookie = (key: string) => {
      const cookies = document.cookie.split("; ");

      return cookies.find((c) => c.startsWith(key))?.split("=")[1];
    };

    expect(auth.id).toEqual(getCookie("auth-id"));
    expect(auth.token).toEqual(getCookie("auth-token"));

    expect(auth).toMatchSnapshot();
  });
});

interface ResType {
  id: `${number}`;
  token: string;
}

interface Salt {
  salt: string;
}
