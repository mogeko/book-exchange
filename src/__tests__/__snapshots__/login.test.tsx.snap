// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorize API return a hmac salt 1`] = `"w4uh61mm79nxt7re"`;

exports[`Authorize API return an authorization token 1`] = `
Object {
  "id": "90019",
  "token": "2d720993-4aed-43c6-a653-90d3f9a87e9b",
}
`;
